go/ai/generator_test.go

```diff
@@ -0,0 +1,193 @@
+// limitations under the License.
+
+package ai
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestValidCandidate(t *testing.T) {
+	t.Parallel()
+
+	t.Run("Valid candidate with text format", func(t *testing.T) {
+		candidate := &Candidate{
+			Message: &Message{
+				Content: []*Part{
+					NewTextPart("Hello, World!"),
+				},
+			},
+		}
+		outputSchema := &GenerateRequestOutput{
+			Format: OutputFormatText,
+		}
+		_, err := validCandidate(candidate, outputSchema)
+		assert.NoError(t, err)
+	})
+
+	t.Run("Valid candidate with JSON format and matching schema", func(t *testing.T) {
+		json := `{
+			"name": "John",
+			"age": 30,
+			"address": {
+				"street": "123 Main St",
+				"city": "New York",
+				"country": "USA"
+			}
+		}`
+		candidate := &Candidate{
+			Message: &Message{
+				Content: []*Part{
+					NewTextPart(JSONMarkdown(json)),
+				},
+			},
+		}
+		outputSchema := &GenerateRequestOutput{
+			Format: OutputFormatJSON,
+			Schema: map[string]any{
+				"type":     "object",
+				"required": []string{"name", "age", "address"},
+				"properties": map[string]any{
+					"name": map[string]any{"type": "string"},
+					"age":  map[string]any{"type": "integer"},
+					"address": map[string]any{
+						"type":     "object",
+						"required": []string{"street", "city", "country"},
+						"properties": map[string]any{
+							"street":  map[string]any{"type": "string"},
+							"city":    map[string]any{"type": "string"},
+							"country": map[string]any{"type": "string"},
+						},
+					},
+					"phone": map[string]any{"type": "string"},
+				},
+			},
+		}
+		candidate, err := validCandidate(candidate, outputSchema)
+		assert.NoError(t, err)
+		text, err := candidate.Text()
+		assert.NoError(t, err)
+		assert.EqualValues(t, text, json)
+	})
+
+	t.Run("Invalid candidate with JSON format and non-matching schema", func(t *testing.T) {
+		candidate := &Candidate{
+			Message: &Message{
+				Content: []*Part{
+					NewTextPart(JSONMarkdown(`{"name": "John", "age": "30"}`)),
+				},
+			},
+		}
+		outputSchema := &GenerateRequestOutput{
+			Format: OutputFormatJSON,
+			Schema: map[string]any{
+				"type": "object",
+				"properties": map[string]any{
+					"name": map[string]any{"type": "string"},
+					"age":  map[string]any{"type": "integer"},
+				},
+			},
+		}
+		_, err := validCandidate(candidate, outputSchema)
+		assert.Error(t, err)
+		assert.Contains(t, err.Error(), "data did not match expected schema")
+	})
+
+	t.Run("Candidate with invalid JSON", func(t *testing.T) {
+		candidate := &Candidate{
+			Message: &Message{
+				Content: []*Part{
+					NewTextPart(JSONMarkdown(`{"name": "John", "age": 30`)), // Missing trailing }.
+				},
+			},
+		}
+		outputSchema := &GenerateRequestOutput{
+			Format: OutputFormatJSON,
+		}
+		_, err := validCandidate(candidate, outputSchema)
+		assert.Error(t, err)
+		assert.Contains(t, err.Error(), "data is not valid JSON")
```

Comment:
I believe this will panic if err == nil. See my comment above.

Reply:
Just to be clear, it is fine to write helper functions to make it easier to write tests. But they are best kept local.
For example, you have a lot of tests that look for strings in the error message. (Which we also frown upon, but that's another story.) You could write a helper:
```
func errorContains(t *testing.T, err error, want string) {
   t.Helper()
   if err == nil {
       t.Error("got nil, want error")
   } else if !strings.Contains(err.Error(), want) {
      t.Errorf("got error message %q, want it to  contain %q", err, want)
   }
}
```

Reply:
Used the helper here, thanks!